kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/OauthAuthorityExtractor.java

```diff
@@ -25,7 +47,64 @@ public Mono<Set<String>> extract(AccessControlService acs, Object value, Map<Str
       throw new RuntimeException();
     }
 
-    return Mono.just(Set.of(principal.getName())); // TODO #2844
+    var provider = (OAuthProperties.OAuth2Provider) additionalParams.get("provider");
+    var rolesFieldName = provider.getCustomParams().get(ROLES_FIELD_PARAM_NAME);
+
+    Set<String> rolesByUsername = acs.getRoles()
+        .stream()
+        .filter(r -> r.getSubjects()
+            .stream()
+            .filter(s -> s.getProvider().equals(Provider.OAUTH))
+            .filter(s -> s.getType().equals("user"))
+            .anyMatch(s -> s.getValue().equals(principal.getName())))
+        .map(Role::getName)
+        .collect(Collectors.toSet());
+
+    Set<String> rolesByRolesField = acs.getRoles()
+        .stream()
+        .filter(role -> role.getSubjects()
+            .stream()
+            .filter(s -> s.getProvider().equals(Provider.OAUTH))
+            .filter(s -> s.getType().equals("role"))
+            .anyMatch(subject
+                -> {
+              var principalRoles = convertRoles(principal.getAttribute(rolesFieldName));
+              var roleName = subject.getValue();
+              return principalRoles.contains(roleName);
+            })
+        )
+        .map(Role::getName)
+        .collect(Collectors.toSet());
+
+    return Mono.just(Stream.concat(rolesByUsername.stream(), rolesByRolesField.stream()).collect(Collectors.toSet()));
+  }
+
+  @SuppressWarnings("unchecked")
+  private Collection<String> convertRoles(Object roles) {
+    if (roles == null) {
+      log.debug("Param missing from attributes, skipping");
+      return Collections.emptySet();
+    }
+
+    try {
+      if ((roles instanceof List<?>) || (roles instanceof Set<?>)) {
+        log.trace("The field is either a set or a list, returning as is");
+        return (Collection<String>) roles;
+      }
+
+      if (!(roles instanceof String)) {
+        log.debug("The field is not a string, skipping");
+        return Collections.emptySet();
+      }
+
+      log.trace("Trying to deserialize the field");
+      //@formatter:off
+      return new ObjectMapper().readValue((String) roles, new TypeReference<>() {});
```

Comment:
nit: I think re-creating `ObjectMapper` for every request is a bit an overhead. Probably we can make it a class field which is populated as a constant or from outside through the constructor. Please, share your thoughts.